### PR TITLE
Add function "hess" for Hessenberg decomposition

### PR DESCRIPTION
**Pull request by sh-zheng**
Hi~ This patch add new function "hess" and realize the upper hessenberg decomposition. I have test it for single/double precision and real/complex matrix with openblas lab.

### CODE REVIEW

File: include/armadillo_bits/def_lapack.hpp [hunk 1]
```diff
@@ -23,7 +23,11 @@
 #endif
 
 #if !defined(ARMA_BLAS_CAPITALS)
-  
+  #define arma_sgehrd sgehrd
```
Comment (conradsnicta) [inline include/armadillo_bits/def_lapack.hpp]: the uppercase versions of the lapack names are also needed; see the defines after the #else statement

File: include/armadillo_bits/auxlib_meat.hpp [hunk 2]
```diff
@@ -1883,6 +1883,56 @@ auxlib::chol_band_common(Mat<eT>& X, const uword KD, const uword layout)
...
+      {
+      H.reset();
+      return true;
+      }
+    
+    arma_debug_assert_blas_size(H);
+    
+    const uword H_n_rows = H.n_rows;
+    
+    blas_int  n      = H_n_rows;
```
Comment (conradsnicta) [inline include/armadillo_bits/auxlib_meat.hpp]: `H_n_rows` must be cast to a `blas_int` before assignment to `n`.
For example: `blas_int n = blas_int(H_n_rows);`

File: include/armadillo_bits/fn_hess.hpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+  const typename arma_blas_type_only<typename T1::elem_type>::result* junk = 0
+  )
+  {
+  arma_extra_debug_sigprint();
+  arma_ignore(junk);
+  
+  typedef typename T1::elem_type eT;
+  
+  Mat<eT> H;
+  Col<eT> tao(X.get_ref().n_rows-1);
```
Comment (conradsnicta) [inline include/armadillo_bits/fn_hess.hpp]: Same problems as described for the earlier function.

File: include/armadillo_bits/fn_hess.hpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  
+  const bool status = auxlib::hess(H, X.get_ref(), tao);
+  for (int i=0; i<H.n_rows-2; i++){
```
Comment (conradsnicta) [inline include/armadillo_bits/fn_hess.hpp]: Same problems as described for the earlier function.

File: include/armadillo_bits/fn_hess.hpp [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+  const typename arma_blas_type_only<typename T1::elem_type>::result* junk = 0
+  )
+  {
+  arma_extra_debug_sigprint();
+  arma_ignore(junk);
+  
+  arma_debug_check( void_ptr(&U) == void_ptr(&H), "hess(): 'U' is an alias of 'H'" );
+  
+  typedef typename T1::elem_type eT;
+  Col<eT> tao(X.get_ref().n_rows-1);
```
Comment (conradsnicta) [inline include/armadillo_bits/fn_hess.hpp]: Same problems as described for the earlier function.

**Comment by conradsnicta** [comment]
This pull request currently has too many problems.

- instead of `int`, use `uword`
- take into account possible aliasing
- don't assume that the input to the function is a matrix; it could be an expression like A+B
- take into account that the expression can evaluate to an empty matrix (eg. 0 rows), or a matrix with only one row; the current code uses arithmetic which will cause overflows in these cases.
- take into account that users can have a version of LAPACK where all the function names are in uppercase
- be careful when doing blind assignment to `blas_int`; each value must first be cast to `blas_int`

**Comment by sh-zheng** [comment]
Sorry, I am careless and too eager to submit them after the test passed : ). Thank you very much for checking these codes, and I will fix the bugs.

**Pull request closed without merging**